### ISSUE TITLE
CppSharp does not properly support new GCC5 ABI

### ISSUE THREAD

**Issue by cometyang** [mono/CppSharp#726]
##### Brief Description; "

When build cppsharp under unbuntu, got the following message:

Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.

OS: Windows / OS X / Linux (include version and/or distro)
ubuntu 16.04

##### Used settings

Target: MSVC/GCC/Clang

Other settings 

##### Stack trace or incompilable generated code

Running prebuild commands
mono --debug "../lib/Release_x64/Encodings.Gen.exe"

Generating bindings for Encodings (CSharp)
Looking for tests in: /home/user1/projects/csharp/CppSharp/tests/Encodings
Stacktrace:

  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) CppSharp.Parser.ClangParser/__Internal.ParseHeader_0 (intptr) <IL 0x00009, 0x0005f>
....

**Comment by tritao** [comment]
This might be related to https://github.com/mono/CppSharp/issues/655.

**Comment by cometyang** [comment]
The error happens when build tests.
I enable verbose mode, and the following is an example:
 g++    -MMD -MP -DNDEBUG -D_GLIBCXX_USE_CXX11_ABI=0 -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -I../../scripts/llvm-cccdd2-linux-Release/include -I../../scripts/llvm-cccdd2-linux-Release/tools/clang/include -I../../scripts/llvm-cccdd2-linux-Release/tools/clang/lib -I../../scripts/llvm-cccdd2-linux-Release/build/include -I../../scripts/llvm-cccdd2-linux-Release/build/tools/clang/include -m64 -O2 -fPIC -g -fpermissive -std=c++11 -fno-rtti  -o "../obj/Release_x64/x64/Release/CppSharp.CppParser/Target.o" -MF "../obj/Release_x64/x64/Release/CppSharp.CppParser/Target.d" -c "../../../src/CppParser/Target.cpp"
Linking CppSharp.CppParser


It seems the flag passed correctly.

**Comment by tritao** [comment]
You can try to change the triple used by the parser bindings to the GNU C++11 one (https://github.com/mono/CppSharp/blob/master/src/CppParser/Bindings/CSharp/premake5.lua#L30), as explained in https://github.com/mono/CppSharp/issues/655#issuecomment-230582585.

**Comment by cometyang** [comment]
I tried all combinations, it does not work.

**Comment by tritao** [comment]
Can reproduce in Travis CI if I update GCC to version 6.

https://travis-ci.org/mono/CppSharp/jobs/203052175

**Comment by tritao** [comment]
I just realized this is probably related to our pre-compiled version of LLVM that is not compatible with GCC 5+.

**Comment by damageboy** [comment]
Can you please document what/how needs to be done to build this on machines that default to gcc 5/6?

For example, trying to build on ubuntu 16.04 blows up with native stack traces...

Should I try and re-run with `CC=gcc-4.9 CXX=g++-4.9` ?
Will this be enough?

This sort of information really belongs in the wiki.

I'll gladly fix the docs and submit a PR. but would like to know what's the proper way of dealing with this...

**Comment by tritao** [comment]
I think what needs to be done here is:

1) Compile LLVM dependencies with GCC6 and create a new package. This can be done with our `LLVM.lua` script, check the guide.

2) Re-generate the parser bindings with `Parser.Gen.exe` (this needs to be done with an already-working version to bootstrap the GCC5 version, so you'll need to compile it with GCC4).

3) Update whatever is needed on Premake scripts so we choose the GCC5 parser bindings in case the system compiler is >= GCC5.

If you need any help with any of it feel free to come to the Gitter channel and we'll help.

**Comment by damageboy** [comment]
Will try this now.
Is this true for any version of GCC? Or only 5/6?

Should forcing gcc with 
```
export CC=gcc-4.9
export CXX=g++-4.9
```

be enough? I'm referring the the step in your comment (2) where you say to run `Parser.Gen.exe`  with already working version / /GCC 4)

**Comment by tritao** [comment]
That should work yes.

**Comment by rokups** [comment]
I believe i am hitting this issue as well. Using `x86_64-linux-gnu-cxx11abi`, gcc7.2. I also have a local llvm build and regenerated c# bindings which changed very little actually (https://github.com/rokups/CppSharp/commit/e3910ae77af1f521e2da4addcbecd4d1c1e39c6a).

Backtraces a little bit descriptive.

Managed:
```
~/src/open/CppSharp/bin/Debug % mono Urho3DNet.exe /tmp/Urho3D-Upstream/include
Parsing libraries...
Parsing code...
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
Stacktrace:

  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) Std.BasicStringExtensions/__Internal.BasicString (intptr,string,intptr) [0x0000b] in <84530c2505554c098560c5764027f607>:0
  at Std.BasicStringExtensions.BasicString (string,Std.Allocator`1<sbyte>) [0x00028] in <84530c2505554c098560c5764027f607>:0
  at CppSharp.Parser.CppParserOptions.set_TargetTriple (string) [0x00007] in <84530c2505554c098560c5764027f607>:0
  at CppSharp.Parser.ParserOptions.BuildForSourceFile (System.Collections.Generic.IEnumerable`1<CppSharp.AST.Module>,string) [0x00034] in <e0a6341113fa49d19c8e19c6dcc4287b>:0
  at CppSharp.Driver.ParseCode () [0x00123] in <5d64dcf4ea2a45a8854dc43f47f38e41>:0
  at CppSharp.ConsoleDriver.Run (CppSharp.ILibrary) [0x000a1] in <5d64dcf4ea2a45a8854dc43f47f38e41>:0
  at Urho3DNet.Program.Main (string[]) [0x0002b] in <a63f645887ff4d88bd7045cf7f4f78be>:0
  at (wrapper runtime-invoke) <Module>.runtime_invoke_void_object (object,intptr,intptr,intptr) [0x0004e] in <a63f645887ff4d88bd7045cf7f4f78be>:0
```
Native:
```Thread 1 (Thread 0x7fe02c361740 (LWP 2688)):
#0  0x00007fe02b823687 in waitpid () from /usr/lib/libpthread.so.0
#1  0x00005589131bd0d5 in ?? ()
#2  <signal handler called>
#3  0x00007fe02b2778a0 in raise () from /usr/lib/libc.so.6
#4  0x00007fe02b278f09 in abort () from /usr/lib/libc.so.6
#5  0x00007fe027231d77 in __gnu_cxx::__verbose_terminate_handler () at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/vterminate.cc:95
#6  0x00007fe02722f8e6 in __cxxabiv1::__terminate (handler=<optimized out>) at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:47
#7  0x00007fe02722f933 in std::terminate () at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:57
#8  0x00007fe02722fb88 in __cxxabiv1::__cxa_throw (obj=obj@entry=0x558914fc6b70, tinfo=0x7fe02751d098 <typeinfo for std::logic_error>, dest=0x7fe0272454c0 <std::logic_error::~logic_error()>) at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/eh_throw.cc:93
#9  0x00007fe02725a0a3 in std::__throw_logic_error (__s=__s@entry=0x7fe0272e0ba0 "basic_string::_M_construct null not valid") at /build/gcc-multilib/src/gcc/libstdc++-v3/src/c++11/functexcept.cc:66
#10 0x00007fe0272ca6e4 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*> (this=this@entry=0x558914f74f90, __beg=0x0, __end=<optimized out>) at /build/gcc-multilib/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/basic_string.tcc:212
#11 0x00007fe0272ca861 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct_aux<char const*> (__end=<optimized out>, __beg=<optimized out>, this=0x558914f74f90) at /build/gcc-multilib/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/basic_string.h:220
#12 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*> (__end=<optimized out>, __beg=<optimized out>, this=0x558914f74f90) at /build/gcc-multilib/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/basic_string.h:239
#13 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string (this=0x558914f74f90, __s=<optimized out>, __a=...) at /build/gcc-multilib/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/basic_string.h:495
#14 0x0000000041984cc0 in ?? ()
#15 0x00007fe02c2c8400 in ?? ()
#16 0x0000558914f74f90 in ?? ()
#17 0x0000000000000000 in ?? ()
```
Relevant code from `basic_string.tcc`:
```cpp
  template<typename _CharT, typename _Traits, typename _Alloc>
    template<typename _InIterator>
      void
      basic_string<_CharT, _Traits, _Alloc>::
      _M_construct(_InIterator __beg, _InIterator __end,
		   std::forward_iterator_tag)
      {
	// NB: Not required, but considered best practice.
	if (__gnu_cxx::__is_null_pointer(__beg) && __beg != __end)
	  std::__throw_logic_error(__N("basic_string::"
				       "_M_construct null not valid"));
```
Looking at relevant setter in c# bindings i can not really see what is going on. Native code seems to expect  a constructor call with 3 arguments but i do not see that in the setter.
```cs
            public string TargetTriple
            {
                // <...>
                set
                {
                    var __allocator0 = new global::Std.Allocator<sbyte>();
                    var __basicString0 = global::Std.BasicStringExtensions.BasicString(value, __allocator0);
                    ((global::CppSharp.Parser.CppParserOptions.__Internal*) __Instance)->targetTriple = *(global::Std.BasicString.__Internalc__N_std_N___cxx11_S_basic_string__C___N_std_S_char_traits__C___N_std_S_allocator__C*) __basicString0.__Instance;
                }
            }
```

Any ideas?

**Comment by rokups** [comment]
Printed out value of `__s` in `Std.cs` and right before the crash constructor is indeed invoked with a empty string.

**Comment by ddobrev** [comment]
Could you please clarify if it's an empty string or null? This matters. In addition, could you please send us a link to the native API you've been trying to invoke so that we can see the native signature?

**Comment by tritao** [comment]
The native API is our own, it's the parser in CppSharp.

**Comment by rokups** [comment]
Oh sorry, `__s` is indeed `null` upon crash.

**Comment by ddobrev** [comment]
@rokups thank you. I am now going to try reproducing and fixing it. I'll let you know.

**Comment by rokups** [comment]
I did a cowboy thing and replaced `__s` with empty string when it is null. It results in a different crash:
```
  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) CppSharp.Parser.ClangParser/__Internal.ParseHeader (intptr) [0x00002] in <0557539238cb439b9fa9faa3fda8481a>:0
  at CppSharp.Parser.ClangParser.ParseHeader (CppSharp.Parser.CppParserOptions) [0x00012] in <0557539238cb439b9fa9faa3fda8481a>:0
  at CppSharp.ClangParser.ParseSourceFiles (System.Collections.Generic.IEnumerable`1<string>,CppSharp.Parser.ParserOptions) [0x0003c] in <8eb348581da9462bbabb0c0450439636>:0
  at CppSharp.ClangParser.ParseSourceFile (string,CppSharp.Parser.ParserOptions) [0x00001] in <8eb348581da9462bbabb0c0450439636>:0
  at CppSharp.Driver.ParseCode () [0x0012d] in <d2dc9a3aa107427c8254929760f51fc4>:0
  at CppSharp.ConsoleDriver.Run (CppSharp.ILibrary) [0x000a1] in <d2dc9a3aa107427c8254929760f51fc4>:0
  at Urho3DNet.Program.Main (string[]) [0x0002b] in <3a3ac1b80a41481684c347fda5c2a37e>:0
  at (wrapper runtime-invoke) <Module>.runtime_invoke_void_object (object,intptr,intptr,intptr) [0x0004e] in <3a3ac1b80a41481684c347fda5c2a37e>:0
```

@nem0 reported having this same new crash on windows so it must be new issue unrelated to this one.

**Comment by ddobrev** [comment]
@rokups thank you for this test, you have now confirmed this is the issue. You'll have the fix a little later.

**Comment by ddobrev** [comment]
@rokups reproduced on our CI: https://travis-ci.org/mono/CppSharp/jobs/309529018#L2378 . Fixing now...

**Comment by ddobrev** [comment]
@rokups could you please explain how you get null passed in to begin with? I can easily just replace null with empty before invoking the native code but I am not sure this is correct since the user would then read back an empty string instead of null. Deliberately throwing an exception might be the better solution but your code would then crash again. Therefore finding out how you pass a null would be helpful.

**Comment by rokups** [comment]
I did not set `driver.ParserOptions.TargetTripple` and it trickled down to native function call.

**Comment by ddobrev** [comment]
@rokups thank you. Please share your opinion about the better solution considering that the native std::string is simply not allowed to take null. @tritao and I will then discuss the solution.

**Comment by rokups** [comment]
IMHO `NullPointerException` would be a reasonable solution. After all user is setting something to null which can not be null. Beats having entire application crash :)

**Comment by ddobrev** [comment]
@rokups fixed. About the parsing crash you get afterwards, any chance you can debug it, for example on Windows?

**Comment by rokups** [comment]
I did some digging on the that crash i mentioned earlier and it looks odd enough that it may be also ABI related. At least strange nature of it makes me think so. I will post details here. In case you think it is worthy it's own issue please let me know.

```
mono: /home/rk/src/open/CppSharp/src/CppParser/Parser.cpp:4046: CppSharp::CppParser::ParserResult* CppSharp::CppParser::Parser::ParseHeader(const std::vector<std::basic_string<char> >&): Assertion `opts->ASTContext && "Expected a valid ASTContext"' failed.
Stacktrace:

  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) CppSharp.Parser.ClangParser/__Internal.ParseHeader (intptr) [0x00002] in <0557539238cb439b9fa9faa3fda8481a>:0
  at CppSharp.Parser.ClangParser.ParseHeader (CppSharp.Parser.CppParserOptions) [0x00012] in <0557539238cb439b9fa9faa3fda8481a>:0
  at CppSharp.ClangParser.ParseSourceFiles (System.Collections.Generic.IEnumerable`1<string>,CppSharp.Parser.ParserOptions) [0x0003c] in <4f80eec9a8b84edf8b68905f87d137bf>:0
  at CppSharp.ClangParser.ParseSourceFile (string,CppSharp.Parser.ParserOptions) [0x00001] in <4f80eec9a8b84edf8b68905f87d137bf>:0
  at CppSharp.Driver.ParseCode () [0x0012d] in <0d75133a9a1643be8cfe01c04b07f334>:0
  at CppSharp.ConsoleDriver.Run (CppSharp.ILibrary) [0x000a1] in <0d75133a9a1643be8cfe01c04b07f334>:0
  at ImGUI.Program.Main (string[]) [0x00006] in <dcacb1448fdd42f3bc4c002ae19b8b8e>:0
  at (wrapper runtime-invoke) <Module>.runtime_invoke_void_object (object,intptr,intptr,intptr) [0x0004e] in <dcacb1448fdd42f3bc4c002ae19b8b8e>:0
```
```
Thread 1 (Thread 0x7f9d472e1740 (LWP 4930)):
#0  0x00007f9d467a3687 in waitpid () from /usr/lib/libpthread.so.0
#1  0x00005596e2c690d5 in ?? ()
#2  <signal handler called>
#3  0x00007f9d461f78a0 in raise () from /usr/lib/libc.so.6
#4  0x00007f9d461f8f09 in abort () from /usr/lib/libc.so.6
#5  0x00007f9d461f00dc in __assert_fail_base () from /usr/lib/libc.so.6
#6  0x00007f9d461f0153 in __assert_fail () from /usr/lib/libc.so.6
#7  0x00007f9d339342af in CppSharp::CppParser::Parser::ParseHeader (this=0x5596e52faf10, SourceFiles=std::vector of length 1, capacity 1 = {...}) at /home/rk/src/open/CppSharp/src/CppParser/Parser.cpp:4046
#8  0x00007f9d33936758 in CppSharp::CppParser::ClangParser::ParseHeader (Opts=0x5596e52e4130) at /home/rk/src/open/CppSharp/src/CppParser/Parser.cpp:4364
<...>
#39 0x0000000000000000 in ?? ()
```
As you can see it fails on assertion https://github.com/mono/CppSharp/blob/master/src/CppParser/Parser.cpp#L4046 which is invoked by managed code.

I debugged managed application that generates bindings and at position https://github.com/mono/CppSharp/blob/master/src/CppParser/Bindings/CSharp/x86_64-linux-gnu-cxx11abi/CppSharp.CppParser.cs#L19224 i can see `ASTContext` being definitely set.
![1](https://user-images.githubusercontent.com/19151258/33436740-8617de94-d5ee-11e7-9ae2-60199440a224.png)
Restarting process in a native debugger and breaking at https://github.com/mono/CppSharp/blob/master/src/CppParser/Parser.cpp#L4359 i can already see `ASTContext` being null.

**Comment by ddobrev** [comment]
I see, thank you for your work. This is definitely going to be helpful even though I have little idea about it so far.
There's something else I need to ask you. Earlier you said that you had trouble because TargetTriple was null for you. However, this is a native field we declare like this:

    std::string targetTriple;

which means it's not null but rather empty by default. I've just confirmed it in my (Windows) debugger. So it's not normal for it to be null unless you set it. Could you please check how it happens that it's null for you?

**Comment by rokups** [comment]
Getting `TargetTripple` value eventually arrives here https://github.com/mono/CppSharp/blob/master/src/CppParser/Bindings/CSharp/x86_64-linux-gnu-cxx11abi/Std.cs#L1341 and result of that call is null. `__arg0` however is not null.

**Comment by ddobrev** [comment]
It looks like we have a bug in our getter. Let me try a quick test and see if it crashes on Linux.

**Comment by ddobrev** [comment]
I am afraid I cannot reproduce this one on our CI but it's definitely a bug - any not otherwise initialised std::string from C++ must come as an empty string to C#. I am afraid I have some other issues to resolve now. We'll talk more about these problems tomorrow.

**Comment by rokups** [comment]
@ddobrev i have prepared some instructions for you how to reproduce this issue. All you need is linux distribution with docker installed.

In terminal run:
1. `xhost +local:all`
2. `sudo docker run -it --rm -e DISPLAY=$DISPLAY -v/tmp/.X11-unix:/tmp/.X11-unix -v /path/to/cloned/CppSharp:/CppSharp "base/archlinux:latest" bash`

This will spawn you archlinux shell. `--rm` option means container will be removed once you exit shell so you may or may not want to use it.
`xhost +local:all` and `-e DISPLAY=$DISPLAY -v/tmp/.X11-unix:/tmp/.X11-unix` are optional and required only if you wish to run graphical application from within container. 

Inside Archlinux shell:
1. `pacman -Syu --noconfirm`
2. `pacman -S --noconfirm base-devel mono unzip wget nano cmake ninja`
3. `build/InstallNugets.sh`
4. `build/premake5-linux-64 --file=build/scripts/LLVM.lua build_llvm`
5. `build/Compile.sh`

This all is pretty self-explanatory. Before `Compile.sh` exits you will see it crash in `CppSharp.Parser.ClangParser/__Internal.ParseHeader`.

**Comment by rokups** [comment]
I did some more digging. Added `printf("%p\n", opts->ASTContext);` to the start of `ParseHeader()` next to assert. Compiled application and disassembled it. This is what i found:
```
lea     rdi, asc_22E2EA3 ; "%p\n"
mov     rsi, [rax+0C8h]
xor     eax, eax
call    _printf
```
According to sysv ABI on x64 `rsi` register is a second argument, so compiler looks up `ASTContext` at offset 0xC8 (decimal 200). This is wrapper code 
```
[FieldOffset(224)]
internal global::System.IntPtr ASTContext;
```
So... Wrapper has incorrect offset. Obvious thing to try next is to regenerate wrappers in my build. That ofc fails with same error because.. Sort of chicken and egg problem.

**Comment by tritao** [comment]
Did you remove this?

https://github.com/mono/CppSharp/blob/master/src/CppParser/premake5.lua#L26

**Comment by rokups** [comment]
... yup that was it.. It finally works, thank you :)

**Comment by rokups** [comment]
@cometyang you should rest this again. Issue you were having is likely fixed. Just make sure you also build LLVM on your machine.

**Comment by tritao** [comment]
Thanks @rokups!

**Comment by rokups** [comment]
Actually this needs to be reopened. When i claimed that cxx11 build worked i was wrong. I built lib without cxx11 and somehow messed up when checking if its cxx11 or not.

I also tracked down actual problem. With cxx11 we now are getting following crash:
```
Thread 1 (Thread 0x7f73d01ff780 (LWP 30251)):                                                                                                                                                                                                        
#0  0x00007f73cf6c3687 in waitpid () from /usr/lib/libpthread.so.0                                                                                                                                                                                   
#1  0x0000557f03a03580 in ?? ()                                                                                                                                                                                                                      
#2  <signal handler called>                                                                                                                                                                                                                          
#3  0x00007f73cf117860 in raise () from /usr/lib/libc.so.6                                                                                                                                                                                           
#4  0x00007f73cf118ec9 in abort () from /usr/lib/libc.so.6                                                                                                                                                                                           
#5  0x00007f73cf15a437 in __libc_message () from /usr/lib/libc.so.6                                                                                                                                                                                  
#6  0x00007f73cf160d34 in malloc_printerr () from /usr/lib/libc.so.6                                                                                                                                                                                 
#7  0x00007f73cf162439 in _int_free () from /usr/lib/libc.so.6                                                                                                                                                                                       
#8  0x00007f73c541ec82 in __gnu_cxx::new_allocator<char>::deallocate (this=0x7ffc31eea6a0, __p=<optimized out>) at /usr/include/c++/7.2.1/ext/new_allocator.h:125                                                                                    
#9  std::allocator_traits<std::allocator<char> >::deallocate (__a=..., __n=<optimized out>, __p=<optimized out>) at /usr/include/c++/7.2.1/bits/alloc_traits.h:462                                                                                   
#10 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_destroy (__size=<optimized out>, this=0x7ffc31eea6a0) at /usr/include/c++/7.2.1/bits/basic_string.h:226                                                      
#11 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose (this=0x7ffc31eea6a0) at /usr/include/c++/7.2.1/bits/basic_string.h:221                                                                              
#12 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string (this=0x7ffc31eea6a0, __in_chrg=<optimized out>) at /usr/include/c++/7.2.1/bits/basic_string.h:647                                                
#13 CppSharp::CppParser::Parser::Setup (this=this@entry=0x557f0507e580) at ../../../src/CppParser/Parser.cpp:268                                                                                                                                     
#14 0x00007f73c542fd11 in CppSharp::CppParser::Parser::ParseHeader (this=0x557f0507e580, SourceFiles=std::vector of length 3, capacity 4 = {...}) at ../../../src/CppParser/Parser.cpp:4058
```
Which stems from following c++ code in `Prser.cpp`:
```cpp
    if (opts->targetTriple.empty())
        opts->targetTriple = llvm::sys::getDefaultTargetTriple();
```
It crashes because `libCppSharp.CppParser.so` tries to free address that was not malloc'ed. See relevant assembly:
```
.text:00007F3F216A9C2D                 mov     r15, [rbp+this]
.text:00007F3F216A9C34                 mov     [r15+18h], edx
.text:00007F3F216A9C38                 mov     rdx, [rax+0F8h] ; opts->targetTriple->size
.text:00007F3F216A9C3F                 test    rdx, rdx
.text:00007F3F216A9C42                 jnz     short loc_7F3F216A9C94
.text:00007F3F216A9C44                 lea     r12, [rbp+return_value] ; return value of next call
.text:00007F3F216A9C4B                 mov     rdi, r12        ; 1st argument
.text:00007F3F216A9C4E                 call    __ZN4llvm3sys22getDefaultTargetTripleB5cxx11Ev ; llvm::sys::getDefaultTargetTriple();
.text:00007F3F216A9C53                 mov     rax, [r15+8]    ; opts
.text:00007F3F216A9C57                 mov     rsi, r12        ; 2nd argument
.text:00007F3F216A9C5A                 add     r12, 10h        ; &ret + 0x10, ???, becomes 00007FFEAF37DA60
.text:00007F3F216A9C5E                 lea     rdi, [rax+0F0h] ; 1st argument, opts->targetTriple
.text:00007F3F216A9C65                 mov     [rbp+var_810], rax ; destination is 00007FFEAF37D950
.text:00007F3F216A9C6C                 call    std::__cxx11::basic_string<char,std::char_traits<char>,std::allocator<char>>::operator=(std::__cxx11::basic_string<char,std::char_traits<char>,std::allocator<char>>&&)
.text:00007F3F216A9C71                 mov     rdi, [rbp+return_value] ; void *
.text:00007F3F216A9C78                 cmp     rdi, r12        ; &ret + 0x10 == ret->(0x00)buffer, it will always be false
.text:00007F3F216A9C7B                 jz      short loc_7F3F216A9C82
.text:00007F3F216A9C7D                 call    operator delete(void *)    ; crash
```
Basically compiler inlines `delete[] opts->targetTriple->buffer`  call here.

Is it possible that internal string buffer is allocated by managed code, say through `AllocHGlobal()` or something? It does not make much sense to me but it is the only thing i can think of..

I tried adding a workaround in this spot, appending to `targetTriple` instead of assigning to it. It works but application then crashes in `CppParserOptions::~CppParserOptions`. There definitely is a serious memory management problem here somewhere.
